Added ORKWeightAnswerFormat

ninoguba opened the pull request:
This adds an answer format for collecting weight mostly based on @rsanchezsaez's ORKHeightAnswerFormat.

rsanchezsaez commented:
Hey @ninoguba, this looks pretty great, thanks for your contribution!

I hope to find some time to careful review the code later, but some initial thoughts:

- It's great that you added a two-slider picker with a detailed `oz` selection, but personally I'd try to mimic the *Apple Health Medical ID* weight picker (see screenshots below, you can check it out on the iOS *Health* app), which only provides `lbs`. Since you already did this and it's nice to have, how about we provide an option when constructing the weight picker to chose between *lbs* only (which would be the default) or *lbs and oz*? Since this is locale dependent, I'd name the option `(BOOL)additionalPrecision`. `NO` would show *lbs* only, `YES` would show both. 
- Please also note that there are *16 oz* in a *lb*.
- Same for the `kgs` picker, for it to mimic the *Apple Health Medical ID* one: I'd just use a one-slider picker, with *.5* increments. `additionalPrecision` would show a single column *kgs* picker with *.1* increments.
- Also, I think that the upper limit and default selections should be the same as on the *Apple Health Medical ID* ones (certainly there are people who weight more than *150 kg*) ;-).

What do you think?

---

Note to self: I think the default selections on the height picker changed from `5ft 4in` to `5ft 6in`, I'll need to update those too.

---

![img_1887](https://user-images.githubusercontent.com/444313/30414570-1dc93b00-98d9-11e7-8543-08feb1c3949a.png)
![img_1889](https://user-images.githubusercontent.com/444313/30414575-1decbeb8-98d9-11e7-8c6d-ac251c2321dc.png)
![img_1888](https://user-images.githubusercontent.com/444313/30414573-1de6f578-98d9-11e7-8052-580ac12306ba.png)
![img_1890](https://user-images.githubusercontent.com/444313/30414572-1ddc20e4-98d9-11e7-8616-6f43268feb29.png)
![img_1891](https://user-images.githubusercontent.com/444313/30414571-1ddbd9f4-98d9-11e7-912e-68c98cffde1f.png)
![img_1892](https://user-images.githubusercontent.com/444313/30414574-1de8d12c-98d9-11e7-834a-3c58628fadbe.png)

---

ninoguba commented:
@rsanchezsaez thanks for the feedback.  I agree with all your suggestions and will work on them.

Have any thoughts on the default value for weight?  Medical ID uses 144 lbs, is this based on some universally acceptable average weight?  I think I'll just add an initializer that developers can specify their desired default/min/max values. -->  I think this would be great for your ORKHeightAnswerFormat too, for flexibility and consistency.

ninoguba commented:
@rsanchezsaez ok done with what you've suggested except I kept the metric weight picker as kg whole numbers only (if additionalPrecision=NO) to keep it simple.  If anyone needs the fractions then they can with additionalPrecision=YES.

I've also corrected the conversion of pounds and ounces.  Thanks for catching that!

And for added flexibility, and validation in a way, I have added parameters for instantiating it with default/min/max values.

Please review when you get the chance.

ninoguba commented:
This PR is also manifesting bug #1012 so working on the fix...

ninoguba commented:
Bug in new weight answer format similar to the one described in #1012 fixed.

rsanchezsaez commented:
Thanks for the changes!

I still think that `additionalPrecision=NO` should default to the same as used by Apple on the Medical ID Picker: .5 increments for kgs, for coherency. :-)

ninoguba commented:
I just find it oddly inconsistent when additionalPrecision=NO the USC system will be showing pounds in whole number values while the metric system will be showing kilograms in .5 increments.  This PR would satisfy those who wouldn't want to show any decimal places at all.

ninoguba commented:
Sharing the screenshots...

![simulator screen shot sep 20 2017 9 35 32 am](https://user-images.githubusercontent.com/4361804/30655930-37160dc8-9de7-11e7-8368-9186d3422d98.png)
![simulator screen shot sep 20 2017 9 35 38 am](https://user-images.githubusercontent.com/4361804/30655929-3715dd30-9de7-11e7-81ad-5ada52fad37f.png)
![simulator screen shot sep 20 2017 9 35 45 am](https://user-images.githubusercontent.com/4361804/30655927-371289e6-9de7-11e7-86e8-28ab611a8924.png)
![simulator screen shot sep 20 2017 9 35 50 am](https://user-images.githubusercontent.com/4361804/30655928-3712a278-9de7-11e7-8a35-bf3b255a80dc.png)

rsanchezsaez commented:
I insist in thinking that the defaults should completely mimic in look, feel, and unit increment the Medical ID ones, even though that feels slightly non-coherent from the developer perspective.

I think maintaining user experience consistency is more important than maintaining developer consistency. It’s unlikely that a single user will see both locales, but on the other hand it is probable that a ResearhKit user answering this step has already seen the Medical ID picker. ;-)

I think Apple has sensible reasons for choosing these defaults and I’d just adopt them here too. That said, we can make them even more configurable if we want, or rename the argument name.

ninoguba commented:
I could add a valueInterval attribute similar to UIDatePicker's [minuteInterval](https://developer.apple.com/documentation/uikit/uidatepicker/1615998-minuteinterval) attribute.  

For metric and additionalPrecision=NO, the default will be 5, min will be 1, max will be 5, and value should be evenly divided into 10.  

For additionalPrecision=YES and USC, this valueInterval attribute will be ignored.

What do you think?

ninoguba commented:
@rsanchezsaez done, the default metric weight picker now matches that of Medical ID.

![simulator screen shot sep 20 2017 9 08 12 pm](https://user-images.githubusercontent.com/4361804/30678582-13cfcf82-9e48-11e7-9fff-559fc6030697.png)

To maintain developer flexibility, I added the valueInterval attribute that can be set to 0.0 if they want to present it with no decimals.

![simulator screen shot sep 20 2017 9 08 17 pm](https://user-images.githubusercontent.com/4361804/30678632-64e39d40-9e48-11e7-964b-f8c8b74a9fb2.png)

ninoguba commented:
@rsanchezsaez @Pariecemckinney-apple good for merge?

akshay-yadav-apple commented:
Hey @ninoguba,
I reviewed the PR, and it looks good and ready to merge in. 
@Pariecemckinney-apple, @rsanchezsaez any thoughts/suggestions before merging?

rsanchezsaez commented on ResearchKit/Common/ORKAnswerFormat.h:
```diff
@@ -1431,6 +1441,104 @@ ORK_CLASS_AVAILABLE
+ 
+ The default is .5 otherwise zero. If value is non-zero or if 'nil' the default is used.
+ */
+@property (copy, readonly, nullable) NSNumber *valueInterval;
```
If I understood correctly, it seems that it currently is only used as a boolean flag for choosing between 0.5kg and 1kg increments in metric (it is ignored in USC), and the actual numeric `valueInterval` is not used anywhere. Is that right?

Either:
- Consider converting to a `BOOL` named something like `useWholeMetricValues` (easier).
- Consider implementing proper metric value intervals (renaming to `metricValueInterval`) (Maybe it's not worth the effort, as I don't think users will find much benefit in being able to fine tune metric precision that much).
- Consider unifying `valueInterval` and `additionalPrecision` into a general `precision` `NS_ENUM` mask which accepts the following values
  1. `ORKPrecisionSimple` (1 kg increments / 1 lb increments with no ounces).
  2. `ORKPrecisionDefault` (0.5 kg increments / 1 lb increments with no ounces).
  3. `ORKPrecisionAdditional` (1 gram increments / lb and ounces).

(This is my preferred option, as it condenses all currently available options into a neat single enum, provides maximum flexibility, and keeps the defaults in line with iOS's on both measurement systems).

What do you think?

Let me know if you need help doing this and I'll be happy to do this part this week.

ninoguba replied:
You are correct.  I ended up using it only as a boolean flag.  Using enums would definitely make things much cleaner.

rsanchezsaez commented on ResearchKit/Common/ORKAnswerFormat.h:
```diff
@@ -1431,6 +1441,104 @@ ORK_CLASS_AVAILABLE
+@property (copy, readonly, nullable) NSNumber *maximumValue;
+
+/**
+ The interval at which the picker should display kilograms. (ignored for USC and additionalPrecision=YES)
```
(Ignore for now if we want to do my suggested unification below.)

rsanchezsaez commented on ResearchKit/Common/ORKAnswerFormat.h:
```diff
@@ -1431,6 +1441,104 @@ ORK_CLASS_AVAILABLE
+/**
+ A Boolean value indicating whether the value picker will display additional precision. (read-only)
+
+ By default, the value of this property is `NO`.
```
(Ignore for now if we want to do my suggested unification above.)

Pariecemckinney-apple commented:
@akshay-yadav-apple Looks good!, and thank you @rsanchezsaez for the thorough review! Thanks @ninoguba !

ninoguba commented:
@rsanchezsaez very thorough indeed, thanks!  so am I to implement these suggested changes first prior to being merged?   Or merge first then let @rsanchezsaez do all his suggested edits to credit his great contribution to this PR.  It's his ORKHeightAnswerFormat that is the basis of this one anyway so I wont mind at all. =)

rsanchezsaez commented:
@ninoguba Whatever you prefer, but from the commit history perspective it'd be a bit cleaner if we do it before merging this PR.

I don't mind at all you doing them if you have the time. And I don't mind helping out if you don't have the time (in that case, I can submit a secondary-PR against your own branch, and then if you merge it, my commits would get automatically added to your PR before it being merged into the main RK `master` branch).

akshay-yadav-apple commented:
I agree with @rsanchezsaez. It'd be best to add the changes before merging the PR.

ninoguba commented:
@rsanchezsaez lets do what you suggested --> "I can submit a secondary-PR against your own branch, and then if you merge it, my commits would get automatically added to your PR before it being merged into the main RK master branch"

ninoguba commented on ResearchKit/Common/ORKAnswerFormat.m:
```diff
@@ -2731,6 +2754,132 @@ - (NSString *)stringForAnswer:(id)answer {
+    
+    __typeof(self) castObject = object;
+    return (isParentSame &&
+            (self.measurementSystem == castObject.measurementSystem));
```
Yes agreed.

ninoguba commented on ResearchKit/Common/ORKAnswerFormat.m:
```diff
@@ -2731,6 +2754,132 @@ - (NSString *)stringForAnswer:(id)answer {
+    
+    if (!ORKIsAnswerEmpty(answer)) {
+        NSNumberFormatter *formatter = ORKDecimalNumberFormatter();
+        if (self.useMetricSystem) {
```
This was done because the whole and fraction values are separated.  You're welcome to optimize as you see fit.

ninoguba commented on ResearchKit/Common/ORKHelpers_Internal.h:
```diff
@@ -300,6 +300,52 @@ ORK_INLINE double ORKFeetAndInchesToCentimeters(double feet, double inches) {
+
+ORK_INLINE double ORKWholeAndFractionsToKilograms(double whole, double fraction) {
+    double kg = (whole + (fraction / 100));
+    return (round(100*kg)/100);
```
This was done to correct the bug #1012

ninoguba commented on ResearchKit/Localized/en.lproj/ResearchKit.strings:
```diff
@@ -180,6 +180,9 @@
 "MEASURING_UNIT_FT" = "ft";
 "MEASURING_UNIT_IN" = "in";
+"MEASURING_UNIT_KG" = "kg";
+"MEASURING_UNIT_LBS" = "lbs";
```
If changing the default to "lb" then let's provide a way for a developer to override this for cases they need to use "lbs".  This is particularly important for patient-facing apps where the colloquial usage is what's expected by most users.

rsanchezsaez replied:
I kindly disagree.

I wrote *colloquial*, but I meant to say *incorrect*. It's like if Apple were to use slang, or use incorrect English usages such as *your* in place of *you're* on their official iOS translations just because many people do that mistake. I think we should promote correct English within serious apps such ResearchKit-based ones. I think mostly no people will be put off by an app displaying `lb` instead of `lbs`, even if they're more used to the latter. 

See these answers for further clarification: https://english.stackexchange.com/questions/113142/correct-usage-of-lbs-as-in-pounds-of-weight

rsanchezsaez commented:
@ninoguba Thanks for your comments! I've started working on the improvements in the https://github.com/rsanchezsaez/ResearchKit/tree/rss/weight_answer_format branch. I'll have it ready soon.

rsanchezsaez commented:
All my suggested improvements submitted here: https://github.com/ninoguba/ResearchKit/pull/1

ninoguba commented:
@rsanchezsaez ready for the other PR to resolve these conflicts...

ninoguba commented:
@rsanchezsaez thank you!

@akshay-yadav-apple @Pariecemckinney-apple ready to be merged...

akshay-yadav-apple commented:
Wonderful! I will do a quick/final review, and merge.

ninoguba commented on ResearchKit/Common/ORKAnswerFormat.h:
```diff
@@ -160,12 +161,27 @@ ORK_CLASS_AVAILABLE
++ (ORKWeightAnswerFormat *)weightAnswerFormatWithMeasurementSystem:(ORKMeasurementSystem)measurementSystem
+                                                  numericPrecision:(ORKNumericPrecision)numericPrecision;
+
++ (ORKWeightAnswerFormat *)weightAnswerFormatWithMeasurementSystem:(ORKMeasurementSystem)measurementSystem
```
@rsanchezsaez btw, do you think having default, min, max params for the ORKHeightAnswerFormat would be good too?

rsanchezsaez replied:
Yeah, that's not a bad idea. Let's consider adding them in a separate PR. :-)

rsanchezsaez commented:
Note that this PR most probably breaks serialization tests again, but I can easily fix in a separate PR after this PR and https://github.com/ResearchKit/ResearchKit/pull/1031 are merged.

ninoguba commented:
@rsanchezsaez looks like there is one more conflict.

ninoguba commented:
@rsanchezsaez thanks for fixing the conflicts!

@akshay-yadav-apple ready for your review...

akshay-yadav-apple commented:
@ninoguba, unfortunately we merged a PR today which made a number of changes to the framework. Would you mind resolving the small conflict and updating the PR and we can move forward with merging. Thanks, and we look forward to this wonderful addition.

ninoguba commented:
@akshay-yadav-apple conflicts resolved

akshay-yadav-apple commented:
@ninoguba Thanks for resolving the conflicts. 
I am getting some build issues when I tried to build. Can you please confirm if you are able to successfully do a clean build.

ninoguba commented:
Not sure what happened!

ninoguba commented:
@rsanchezsaez help please. I was trying to rebase with master to fix the conflict but looks like I messed up the branch.

ninoguba commented:
ok looks like i'm back to a good state, now let me try and resolve the conflict...

ninoguba commented:
thanks to @rsanchezsaez for helping out fix the conflicts on this PR!

@akshay-yadav-apple I've confirmed the build issues you and I experienced yesterday is not resolved. please review and merge if all ok (before another PR merge conflicts with it again...) thanks!

akshay-yadav-apple commented:
@ninoguba The PR looks good and ready to merge. I am not getting any build issues. Please confirm in the last comment you meant '**...now resolved**'  not '**...not resolved**', and we'll merge the PR.

ninoguba commented:
Sorry for the typo. Yes I meant "now resolved".  I updated it. =) thanks!
